apps/els_lsp/src/els_fungraph.erl

```diff
@@ -0,0 +1,63 @@
+-module(els_fungraph).
+
+-type gnode() :: #{id := _ID, _ => _}.
+-type edges_fun() :: fun((gnode()) -> list(gnode())).
+-opaque graph() :: {?MODULE, edges_fun()}.
+
+-export_type([gnode/0]).
+-export_type([graph/0]).
+
+-export([new/1]).
+-export([traverse/4]).
+
+-spec new(edges_fun()) -> graph().
+new(EdgesFun) ->
+    {?MODULE, EdgesFun}.
+
+-type acc_fun(AccT) :: fun((gnode(), AccT) -> AccT).
+
+-spec traverse(acc_fun(AccT), AccT, _From :: gnode(), graph()) -> AccT.
+traverse(AccFun, Acc, From, G) ->
+  traverse(AccFun, Acc, [From], sets:new(), G).
+
+-spec traverse(acc_fun(AccT), AccT, [gnode()], sets:set(), graph()) ->
+  AccT.
+traverse(AccFun, Acc, [Node | Rest], Visited, G = {?MODULE, EdgesFun}) ->
+  {AdjacentNodes, VisitedNext} = lists:foldr(
+    fun(Adjacent = #{id := ID}, {NodesAcc, VisitedAcc}) ->
+      case sets:is_element(ID, VisitedAcc) of
+        false -> {[Adjacent | NodesAcc], sets:add_element(ID, VisitedAcc)};
+        true -> {NodesAcc, VisitedAcc}
+      end
+    end,
+    {[], Visited},
+    EdgesFun(Node)
+  ),
+  AccNext = lists:foldl(AccFun, Acc, AdjacentNodes),
+  traverse(AccFun, AccNext, AdjacentNodes ++ Rest, VisitedNext, G);
+traverse(_AccFun, Acc, [], _Visited, _G) ->
+  Acc.
+
+-ifdef(TEST).
```

Comment:
Nit: I would prefer the tests to be in their own module, especially in a case like this where all the required API is exported anyway.

Reply:
Fair point. I've addressed it in 69c3a1d.